

<h1 align=center><font size = 5>Regression Models with Keras</font></h1>


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

Part 0: <a href="#item30">Cleaning the Data</a>
    
Part A: <a href="#item31">Build a baseline model</a>  
Part B: <a href="#item32">Normalize the data</a>  
Part C: <a href="#item33">Increate the number of epochs</a>  
Part D: <a href="#item33">Repeat part B with increased Hidden Layers</a>  

</font>
</div>


## Part 0: Downloading and Cleaning the Data


In [15]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

import keras
from keras.models import Sequential
from keras.layers import Dense

In [16]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')


Check the dataset for any missing values.


In [17]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

Data is clean and ready to build the regression model.


#### Split data into predictors and target


In [18]:
concrete_data_columns = concrete_data.columns

X = np.asarray(concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']]) # all columns except Strength (predictors)
y = np.asarray(concrete_data['Strength']) # Strength column (target)


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=3)

X_train.shape

(721, 8)

## Part C: Repeat Part B but use 100 epochs this time for training


 -->

In [19]:
predictors_norm = (X - X.mean()) / X.std()


In [20]:
n_cols = predictors_norm.shape[1] # number of predictors

In [21]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [22]:
# build the model
model = regression_model()

In [23]:
# fit the model
model.fit(X_train, y_train, epochs=, verbose=2)

# Long outputs have been collapsed. Press ... to view them

Epoch 1/50
 - 1s - loss: 1504.1400
Epoch 2/50
 - 0s - loss: 699.9513
Epoch 3/50
 - 0s - loss: 419.1913
Epoch 4/50
 - 0s - loss: 291.5425
Epoch 5/50
 - 0s - loss: 213.3017
Epoch 6/50
 - 0s - loss: 186.5846
Epoch 7/50
 - 0s - loss: 175.5844
Epoch 8/50
 - 0s - loss: 165.7525
Epoch 9/50
 - 0s - loss: 155.9468
Epoch 10/50
 - 0s - loss: 148.6721
Epoch 11/50
 - 0s - loss: 140.2660
Epoch 12/50
 - 0s - loss: 134.9849
Epoch 13/50
 - 0s - loss: 129.5322
Epoch 14/50
 - 0s - loss: 125.1576
Epoch 15/50
 - 0s - loss: 121.0137
Epoch 16/50
 - 0s - loss: 117.9059
Epoch 17/50
 - 0s - loss: 114.3456
Epoch 18/50
 - 0s - loss: 116.7473
Epoch 19/50
 - 0s - loss: 114.8386
Epoch 20/50
 - 0s - loss: 109.8474
Epoch 21/50
 - 0s - loss: 109.3796
Epoch 22/50
 - 0s - loss: 109.7708
Epoch 23/50
 - 0s - loss: 103.9846
Epoch 24/50
 - 0s - loss: 98.9156
Epoch 25/50
 - 0s - loss: 93.1220
Epoch 26/50
 - 0s - loss: 92.2126
Epoch 27/50
 - 0s - loss: 88.2753
Epoch 28/50
 - 0s - loss: 86.7560
Epoch 29/50
 - 0s - loss: 83.4387

In [24]:
#Evaluate the model on the test data

yhat = model.predict(X_test)
Error = model.evaluate(X_test, y_test)
Error

309/309 [==============================] - 0s 322us/step


83.34543798502209

In [25]:
mean_square_error = mean_squared_error(y_test, yhat)
print("MSE is ", mean_square_error)

MSE is  83.34543412949076


In [26]:
# Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors
MSE_List = []
for i in range(50):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=3)
    model.fit(X_train, y_train, epochs=50, verbose=2)
    MSE = model.evaluate(X_test, y_test, verbose=2)
    print("MSE "+str(i+1)+": "+str(MSE))
    yhat = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, yhat)
    MSE_List.append(mean_square_error)
    
MSE_List = np.array(MSE_List)
mean = np.mean(MSE_List)
standard_deviation = np.std(MSE_List)

# Long outputs have been collapsed. Press ... to view them

Epoch 1/50
 - 0s - loss: 72.0314
Epoch 2/50
 - 0s - loss: 71.8864
Epoch 3/50
 - 0s - loss: 79.2591
Epoch 4/50
 - 0s - loss: 75.7324
Epoch 5/50
 - 0s - loss: 74.1256
Epoch 6/50
 - 0s - loss: 71.3899
Epoch 7/50
 - 0s - loss: 70.6398
Epoch 8/50
 - 0s - loss: 73.0757
Epoch 9/50
 - 0s - loss: 72.0559
Epoch 10/50
 - 0s - loss: 71.7366
Epoch 11/50
 - 0s - loss: 70.9728
Epoch 12/50
 - 0s - loss: 70.9594
Epoch 13/50
 - 0s - loss: 72.8717
Epoch 14/50
 - 0s - loss: 70.0889
Epoch 15/50
 - 0s - loss: 69.8075
Epoch 16/50
 - 0s - loss: 71.8305
Epoch 17/50
 - 0s - loss: 69.5322
Epoch 18/50
 - 0s - loss: 71.6759
Epoch 19/50
 - 0s - loss: 71.5454
Epoch 20/50
 - 0s - loss: 69.8816
Epoch 21/50
 - 0s - loss: 71.6589
Epoch 22/50
 - 0s - loss: 70.2134
Epoch 23/50
 - 0s - loss: 71.5899
Epoch 24/50
 - 0s - loss: 69.8249
Epoch 25/50
 - 0s - loss: 71.0963
Epoch 26/50
 - 0s - loss: 70.1732
Epoch 27/50
 - 0s - loss: 68.6720
Epoch 28/50
 - 0s - loss: 68.2460
Epoch 29/50
 - 0s - loss: 68.9146
Epoch 30/50
 - 0s - los

In [27]:
MSE_List = np.array(MSE_List)
standard_deviation = np.std(MSE_List)
print("The 50 MSE values are " + "\n")
for i in range(50):
    print("MSE " + str(i+1) + ": %.2f" % MSE_List[i])
    

print("Mean of 50 times MSE (normalized data): "+str(mean))
print("Standard Deviation: "+str(standard_deviation))
# 50 MSE values are below

The 50 MSE values are 

MSE 1: 82.32
MSE 2: 68.14
MSE 3: 64.06
MSE 4: 56.35
MSE 5: 51.00
MSE 6: 46.40
MSE 7: 45.59
MSE 8: 45.30
MSE 9: 46.48
MSE 10: 46.95
MSE 11: 45.18
MSE 12: 45.80
MSE 13: 44.34
MSE 14: 46.48
MSE 15: 45.43
MSE 16: 43.15
MSE 17: 44.14
MSE 18: 42.95
MSE 19: 45.27
MSE 20: 42.31
MSE 21: 39.86
MSE 22: 40.84
MSE 23: 41.15
MSE 24: 37.99
MSE 25: 37.85
MSE 26: 37.41
MSE 27: 36.25
MSE 28: 38.29
MSE 29: 38.60
MSE 30: 35.54
MSE 31: 36.20
MSE 32: 38.13
MSE 33: 35.87
MSE 34: 34.77
MSE 35: 34.91
MSE 36: 35.38
MSE 37: 34.81
MSE 38: 34.01
MSE 39: 39.08
MSE 40: 33.60
MSE 41: 34.67
MSE 42: 33.83
MSE 43: 33.83
MSE 44: 34.58
MSE 45: 34.07
MSE 46: 38.26
MSE 47: 38.03
MSE 48: 34.77
MSE 49: 36.24
MSE 50: 34.76
Mean of 50 times MSE (normalized data): 41.82388858272235
Standard Deviation: 9.283754037561378


How does the mean of the mean squared errors compare to that from Step B?

The mean of MSE of step D is only 41.82 and is better than Step B (60.93) and is the best of all 4 Parts


<hr>

Combination of OWn Code and 
Copyright © 2019 [IBM Developer Skills Network](https://cognitiveclass.ai?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DL0101EN-SkillsNetwork-20718188&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DL0101EN-SkillsNetwork-20718188&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DL0101EN-SkillsNetwork-20718188&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DL0101EN-SkillsNetwork-20718188&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license?cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DL0101EN-SkillsNetwork-20718188&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ&cm_mmc=Email_Newsletter-_-Developer_Ed%2BTech-_-WW_WW-_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DL0101EN-SkillsNetwork-20718188&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ).
